### ISSUE TITLE
Rollback functionality

### ISSUE THREAD

**Issue by anthturner** [microsoft/AuthJanitor#46]
We pre-test and expect providers to be as honest as possible in that test, but things may still fail. We can't guarantee permissions for some things unless we try and fail, so there's not a good way to necessarily test every variable.

To that end, we should try to include a "rollback" feature in the future to avoid severe failures if something goes wrong mid-cycle.

**Comment by drub0y** [comment]
Yeah, I was going to write something up on how the entire rekeying workflow has the potential to fail (in a couple ways) and leave things in an inconsistent state where things are potentially broken. I think this a big scan of the implementations needs to happen to identify all the possible inconsistent states things could end up in and then have a discussion how to mitigate.

**Comment by anthturner** [comment]
Feature infrastructure is complete as of #101 but individual Providers don't have rollback logic yet.
